src/RBTree.mo

```diff
@@ -30,13 +30,33 @@
 ///
 /// Note:
 /// * Tree operations, such as retrieval, insertion, and removal create `O(log(n))` temporary objects that become garbage.
+///
+/// Credits:
+///
+/// The core of this implementation is derived from Ken Friis Larsen's [RedBlackMap.sml](https://github.com/kfl/mosml/blob/master/src/mosmllib/Redblackmap.sml)
+/// which itself follows:
+///
+/// * Stefan Kahrs, "Red-black trees with types", Journal of Functional Programming, 11(4): 425-432 (2001), (version 1 in web appendix)[http://www.cs.ukc.ac.uk/people/staff/smk/redblack/rb.html].
+
 
 import Debug "Debug";
 import I "Iter";
 import List "List";
 import Nat "Nat";
 import O "Order";
 
+// TODO: RBTree is missing an `unshare` function to reconstitute the object from its `share`ed data
+
+// TODO: a faster, more compact and less indirect representation would be:
+// type Tree<K, V> = {
+//  #red : (Tree<K, V>, K, V, Tree<K, V>);
+//  #black : (Tree<K, V>, K, V, Tree<K, V>);
+//  #leaf
+//};
+// (this inlines the colors into the variant, flattens a tuple, and removes a (now) redundant optin, for considerable heap savings.
```

Comment:
Cool. How big will the savings be, I wonder?  Seems like a lot fewer points per node.

Reply:
-1 word for Colour +1 word for inlining tuple, option removal saves nothing, but we save on all those extra pairs in the heap (-2 words for GC tuple tag plus tuple size, I think) and few indirections to chase so less instructions I hope.